subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+
+static volatile u32_t last_time;
+
+static volatile u64_t idle_cnt;
+static volatile u64_t non_idle_cnt;
+static volatile u64_t sched_cnt;
+
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
+{
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
+#else
+	return thread == _idle_thread;
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
```

Comment:
So accumulate the deltas (which is the same as just the linear time sampled), OK.
Another way is to just count the u32 wraps, then you need less arithmetic here (more efficient) and you can avoid 64bit arith. Then it is up to you if you want to move the 64 bit arith to the presentation printk.

Reply:
Counting u32 wraps could make us lose accuracy, which is important for performance measurements. It is why i decided to use 64 bits.